README.md
# Extended Semantic Components

[![npm version](https://badge.fury.io/js/%40eqworks%2Fextended-semantic-components.svg)](https://www.npmjs.com/package/@eqworks/extended-semantic-components)
[![Build Status](https://travis-ci.com/EQWorks/extended-semantic-components.svg?branch=master)](https://travis-ci.com/EQWorks/extended-semantic-components)
[![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg)](https://github.com/EQWorks/extended-semantic-components/pulls)

This is a collection of components that we've extended based on [Semantic UI React](https://react.semantic-ui.com/) that are commonly used in many of our projects at [EQ Works](https://www.eqworks.com/), such as [ATOM](https://overlord.eqworks.io/), and [Locus](https://console.locus.place/).

You can see the components in action at <https://eqworks.github.io/extended-semantic-components>

### Installation

```shell
$ yarn add @eqworks/extended-semantic-components
# OR
$ npm install @eqworks/extended-semantic-components
```

And to use it within your project

```js
import {
  SemanticDatePicker,
  DataTable,
  Dropzone,
} from '@eqworks/extended-semantic-components'
```


CHANGELOG
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [3.2.0] - 2019-07-25
### Fixed
- `DataTable` fixed to properly handle using `.map` to generate columns (i.e. DataColumn returned by`{columns.map(...)}`)

### Added
- `DataTable` custom styling now available through `cellStyle`, `rowStyle` and `colStyle` props. These props are objects with string keys denoting how to target the cell/row/col (e.g. '-1' for all, '0' for the first row, '1-1' for the second row's second cell) and values with the following structure:
```
'-1': {
	style: {} // inline styling,
	className: 'computedClassName',
	...semanticProps, // e.g. positive: true
	dynamic: ({ value, row, rowIdx, colIdx }) => { function returning a style object, same shape as above },
	conditions: ({ value, row, rowIdx, colIdx }) => { ... function returning true if the above style is to be applied },
}
```
## [3.1.0] - 2019-05-28
### Fixed
- `DataTable` fix datatable default sorting bug
- `DataTable` fix crash on empty object column value

### Added
- `DataTable` Expose `dataKey` in template function callback

## [3.0.1] - 2019-04-08
### Fixed
- `DataTable` Render No columns selected as the placeholder message with empty dataset.
- `DataTable` unable to sort on float value.

## [3.0.0] - 2019-04-03
### Added
- `Dropzone` component added

### Changed
- *[Breaking Change]* `peerDependencies` updated with React `^16.8` for utilization of hooks.

## [2.3.2] - 2019-03-15
### Fixed
- `DataTable` crashing bug with empty dataset. (https://github.com/EQWorks/extended-semantic-components/issues/37)

## [2.3.1] - 2019-03-12
### Added
- `DataTable` customizable search interface

### Fixed
- `DataTable` defaultSortKey not taking effect due to lack of sortType
- `DataTable` string type sort crash on non-string `(a, b)` values.
- `DataTable` string type sort reversed order ('ascending' resulting in 'descending').

## [2.2.0] - 2019-03-01
### Added
- `DataTable` able to download to honour user picked columns.
- `DataTable` Tokenized search feature.
- Zero config `DataTable`.
- Working storybook of `DateRangePicker`.

### Fixed
- `DataTable` A bug caused by doing search while not in the first page.

## [2.1.0] - 2019-01-28
### Added
- Pickable columns for `DataTable`.
- `DataTable` property `perPage` for maximum number of rows per page when pagination is triggered, defaults to its previously hardcoded value 9.
- `DataTable` property `columns` as an alternative of `children` (`DataTable.Column`) for a simpler way to define composable columns through simple Array manipulations. Only one of `children` or `columns` property is allowed.
- `DataTable` property `defaultSortDir` to specify starting sort direction
- `DataTableColumn` properties `sortable` (default is `true`) and `sortType` (default is `string`) added

### Changed
- Zero searchable columns would be used to determine whether to render `DataTable` control section (along with existing ones).
- Yarn based lock.
- sorting previously done by lodash `orderBy`, now changed to custom sort functions of type `date`, `string` and `basic` (number) to allow for more flexible column sorting

### Fixed
- `DataTable`: Search when cell data is `null` or `undefined`.

### Removed
- `search` property for `DataTable` is no longer effective. Use `searchable` for `DataTable.Column` instead. This is not considered to be a breaking change since they were meant to be used in conjunction.

## [2.0.2] - 2019-01-21
### Changed
- Pure Babel based transpile instead of Parcel based build for `npm publish`.

## [2.0.1] - 2019-01-21
### Changed
- Parcel based build.

### Removed
- Storybook demo (will be done through Netlify host soon)

## [2.0.0] - 2019-01-18
### Added
- Initial version of `data-table` component.

### Changed
- Dependencies cleanup and security updates.
- `README.md` improvements and corrections.
- Rename `semantic-date-picker` to `date-picker` for consistency and conciseness.

## [1.0.0] - 2018-07-30
### Added
- Initial version with `semantic-date-picker` component.


stories/data.js
export const largeDataset = [
  {
    'id': 1,
    'first_name': 'Yardley',
    'last_name': 'Mitroshinov',
    'email': 'ymitroshinov0@webeden.co.uk'
  },
  {
    'id': 2,
    'first_name': 'Dorene',
    'last_name': 'Faill',
    'email': 'dfaill1@eepurl.com'
  },
  {
    'id': 3,
    'first_name': 'Nariko',
    'last_name': 'Grelik',
    'email': 'ngrelik2@epa.gov'
  },
  {
    'id': 4,
    'first_name': 'Pail',
    'last_name': 'Wilcockes',
    'email': 'pwilcockes3@hp.com'
  },
  {
    'id': 5,
    'first_name': 'Ambrosi',
    'last_name': 'Foyston',
    'email': 'afoyston4@issuu.com'
  },
  {
    'id': 6,
    'first_name': 'Rosalia',
    'last_name': 'Kennard',
    'email': 'rkennard5@globo.com'
  },
  {
    'id': 7,
    'first_name': 'Cristobal',
    'last_name': 'Du Fray',
    'email': 'cdufray6@techcrunch.com'
  },
  {
    'id': 8,
    'first_name': 'Melloney',
    'last_name': 'Garthland',
    'email': 'mgarthland7@netlog.com'
  },
  {
    'id': 9,
    'first_name': 'Araldo',
    'last_name': 'Paget',
    'email': 'apaget8@ustream.tv'
  },
  {
    'id': 10,
    'first_name': 'Tam',
    'last_name': 'Bodiam',
    'email': 'tbodiam9@facebook.com'
  },
  {
    'id': 11,
    'first_name': 'Madelena',
    'last_name': 'Skittreal',
    'email': 'mskittreala@friendfeed.com'
  },
  {
    'id': 12,
    'first_name': 'Liesa',
    'last_name': 'Keunemann',
    'email': 'lkeunemannb@paypal.com'
  },
  {
    'id': 13,
    'first_name': 'Sher',
    'last_name': "O'Fogarty",
    'email': 'sofogartyc@prweb.com'
  },
  {
    'id': 14,
    'first_name': 'Berny',
    'last_name': 'Shark',
    'email': 'bsharkd@disqus.com'
  },
  {
    'id': 15,
    'first_name': 'Tildy',
    'last_name': 'Holburn',
    'email': 'tholburne@google.com.br'
  },
  {
    'id': 16,
    'first_name': 'Leela',
    'last_name': 'Zorzini',
    'email': 'lzorzinif@g.co'
  },
  {
    'id': 17,
    'first_name': 'Thomasin',
    'last_name': 'Stonhewer',
    'email': 'tstonhewerg@arstechnica.com'
  },
  {
    'id': 18,
    'first_name': 'Gertie',
    'last_name': 'Bissex',
    'email': 'gbissexh@chicagotribune.com'
  },
  {
    'id': 19,
    'first_name': 'Rea',
    'last_name': 'Nudde',
    'email': 'rnuddei@yelp.com'
  },
  {
    'id': 20,
    'first_name': 'Suzi',
    'last_name': 'Itzkin',
    'email': 'sitzkinj@hexun.com'
  },
  {
    'id': 21,
    'first_name': 'Randy',
    'last_name': 'Serginson',
    'email': 'rserginsonk@wikia.com'
  },
  {
    'id': 22,
    'first_name': 'Johna',
    'last_name': 'Kitchingman',
    'email': 'jkitchingmanl@prlog.org'
  },
  {
    'id': 23,
    'first_name': 'Cissiee',
    'last_name': 'Yansons',
    'email': 'cyansonsm@loc.gov'
  },
  {
    'id': 24,
    'first_name': 'Bradan',
    'last_name': 'Dowding',
    'email': 'bdowdingn@dyndns.org'
  },
  {
    'id': 25,
    'first_name': 'Brucie',
    'last_name': 'Pigeon',
    'email': 'bpigeono@seattletimes.com'
  },
  {
    'id': 26,
    'first_name': 'Grant',
    'last_name': 'Dunkerly',
    'email': 'gdunkerlyp@usa.gov'
  },
  {
    'id': 27,
    'first_name': 'Karrie',
    'last_name': 'Houdmont',
    'email': 'khoudmontq@bloomberg.com'
  },
  {
    'id': 28,
    'first_name': 'Gar',
    'last_name': 'MacKeig',
    'email': 'gmackeigr@cafepress.com'
  },
  {
    'id': 29,
    'first_name': 'Inge',
    'last_name': 'Yellowlea',
    'email': 'iyellowleas@twitpic.com'
  },
  {
    'id': 30,
    'first_name': 'Adele',
    'last_name': 'Reding',
    'email': 'aredingt@freewebs.com'
  }
]

export const monsters = [
  { name: 'Godzilla', origin: 'Earth', dob: new Date('1954-11-03').toISOString(), fearsomeness: 1000 },
  { name: 'Ghidorah', origin: '???', dob: '2008-08-08', fearsomeness: 800 },
  { name: 'Lugia', origin: 'Pokémon Universe', dob: '1990-11-23', fearsomeness: 1001 },
  { name: 'Bulbasaur #12', origin: 'Earth', dob: '1990-12-25', fearsomeness: 2333 },
  { name: 'It', origin: '', dob: '', fearsomeness: 'Hey!' },
  { name: 'Cerberus', origin: 0, dob: null, fearsomeness: 'ruff' },
  { name: 'Cthulhu', origin: null, dob: 0, fearsomeness: 100000 },
]
